contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,278 @@
+    /// @param _minSecurityDeposit Minimum security deposit for a service to be eligible to stake.
+    /// @param _stakingRatio Staking ratio: number of seconds per nonce (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    constructor(uint256 _apy, uint256 _minSecurityDeposit, uint256 _stakingRatio, address _serviceRegistry) {
+        // Initial checks
+        if (_apy == 0 || _minSecurityDeposit == 0 || _stakingRatio == 0) {
+            revert ZeroValue();
+        }
+        if (_serviceRegistry == address(0)) {
+            revert ZeroAddress();
+        }
+
+        apy = _apy;
+        minSecurityDeposit = _minSecurityDeposit;
+        stakingRatio = _stakingRatio;
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev Checks token security deposit.
+    /// @param serviceId Service Id.
+    function _checkTokenSecurityDeposit(uint256 serviceId) internal view virtual {}
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal virtual {}
+
+    /// @dev Stakes the service.
+    /// @param serviceId Service Id.
+    function stake(uint256 serviceId) external {
+        // Check the service conditions for staking
+        (, address multisig, , , , , uint8 state) = IService(serviceRegistry).mapServices(serviceId);
+        // The service must be deployed
+        if (state != 4) {
+            revert();
+        }
+        // Check the service security deposit and token, if applicable
+        _checkTokenSecurityDeposit(serviceId);
+
+        // Transfer the service for staking
+        IService(serviceRegistry).transferFrom(msg.sender, address(this), serviceId);
+
+        // ServiceInfo struct will be an empty one since otherwise the transferFrom above would fail
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        sInfo.multisig = multisig;
+        sInfo.owner = msg.sender;
+        sInfo.nonce = IMultisig(multisig).nonce();
+        sInfo.ts = block.timestamp;
+
+        // Add the service Id to the set of staked services
+        setServiceIds.push(serviceId);
+
+        emit ServiceStaked(serviceId, msg.sender);
+    }
+
+    /// @dev Checkpoint to allocate rewards up until a current time.
+    /// @param serviceId Service Id that unstakes, or 0 if the function is called during the deposit of new funds.
+    function _checkpoint(uint256 serviceId) internal returns (uint256 idx) {
+        // Get the service Id set length
+        uint256 size = setServiceIds.length;
+        uint256 lastBalance = balance;
+
+        // If the balance is zero, just bump the timestamp for each of the staking service
+        if (lastBalance == 0) {
+            for (uint256 i = 0; i < size; ++i) {
+                // Set the current timestamp
+                mapServiceInfo[setServiceIds[i]].ts = block.timestamp;
```

Comment:
I don't like this approach. It will be very hard to see when a service started staking (one needs to look at events and reconstruct). I'd prefer we have two fields in `mapServiceInfo`, one for `start` and one for `lastCheckpoint`